tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -0,0 +1,2918 @@
+        op->getLoc(), output_type, op1_equal_in.getResult());
+
+    output = op2_not_op1.getResult();
+  }
+
+  rewriter.replaceOp(op, {output});
+  return success();
+}
+
+LogicalResult ConvertTFLGreaterOp::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_greater_op = cast<TFL::GreaterOp>(op);
+
+  auto input_lhs_type =
+      tfl_greater_op.lhs().getType().dyn_cast<RankedTensorType>();
+  auto input_rhs_type =
+      tfl_greater_op.rhs().getType().dyn_cast<RankedTensorType>();
+  auto output_type =
+      tfl_greater_op.getResult().getType().dyn_cast<RankedTensorType>();
+  // Not a ranked tensor output
+  if (!input_lhs_type || !input_rhs_type || !output_type) return failure();
+
+  bool input_lhs_is_qtype =
+      input_lhs_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+  bool input_rhs_is_qtype =
+      input_rhs_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+
+  if (input_lhs_is_qtype != output_is_qtype ||
+      input_rhs_is_qtype != output_is_qtype) {
+    op->emitOpError(
+        "ConvertTFLGreaterOp: input/output tensor should "
+        "be all quantized or all native");
+    return failure();
+  }
+
+  Value output;
+  if (output_is_qtype) {
+    auto rescale_type =
+        RankedTensorType::get(output_type.getShape(), rewriter.getI32Type());
+    auto input_lhs_qtype = input_lhs_type.getElementType()
+                               .dyn_cast<mlir::quant::UniformQuantizedType>();
+    auto input_rhs_qtype = input_rhs_type.getElementType()
+                               .dyn_cast<mlir::quant::UniformQuantizedType>();
+    auto output_qtype = output_type.getElementType()
+                            .dyn_cast<mlir::quant::UniformQuantizedType>();
+
+    if (input_lhs_qtype.getScale() != input_rhs_qtype.getScale() ||
+        input_lhs_qtype.getZeroPoint() != input_rhs_qtype.getZeroPoint()) {
+      op->emitOpError(
+          "ConvertTFLGreaterOp: input_x and input_y scale/zp "
+          "must be the same");
+      return failure();
+    }
+
+    auto op1_rescale_lhs =
+        buildRescaleToInt32(rewriter, op, tfl_greater_op.lhs(), 1.0f,
+                            input_lhs_qtype.getZeroPoint());
+    auto op2_rescale_rhs =
+        buildRescaleToInt32(rewriter, op, tfl_greater_op.rhs(), 1.0f,
+                            input_rhs_qtype.getZeroPoint());
+    auto op3_greater_op1_op2 = rewriter.create<tosa::GreaterOp>(
+        op->getLoc(), output_type, op1_rescale_lhs, op2_rescale_rhs);
+
+    output = op3_greater_op1_op2.getResult();
+  } else {
+    auto op1_greater_in = rewriter.create<tosa::GreaterOp>(
+        op->getLoc(), output_type, tfl_greater_op.lhs(), tfl_greater_op.rhs());
+
+    output = op1_greater_in.getResult();
+  }
+
+  rewriter.replaceOp(op, {output});
+  return success();
+}
+
+LogicalResult ConvertTFLGreaterEqualOp::matchAndRewrite(
```

Comment:
Nit: I feel that for the quantized case, a utility function could handle most of these with just a switch on the actual comparison. Maybe a TODO

Reply:
Yes we'll address the duplication using a utility function. Marking as TODO.